### ISSUE TITLE
Lib Version for native html

### ISSUE THREAD

**Issue by pnegahdar** [BearToCode/carta#37]
This editor is so cool, kudos! 

I'm not too familiar with svelte, but how hard would it be to compile this into JS and have a pure JS api to bind it to an element. 

```
carta = new Cart()
render(carta, document.getElementById("editor")) 
```

That way it can be used in a react or vanilla project as well?

**Comment by BearToCode** [comment]
Hi @pnegahdar!
It would be really cool to be able to use the editor with other frameworks such as React and Vue. Honestly, I don't have experience with compiling Svelte and using in other frameworks, but quickly looking online seems tricky but possible. For example, I saw that [bytemd](https://github.com/bytedance/bytemd/tree/main), another popular markdown editor, already does so.

I currently do not have time to dedicate to this, but I'll leave the issue open: I may come back to this in the future, or maybe you or some other developer may help 😄

**Comment by mithi** [comment]
Hi @BearToCode , 

I know that currently you don't have time to dedicate to this, 

I was just wondering if it is possible to `sponsor` you financially or for you to take a `commission` (Not soon maybe 1 to 2 months from now when Svelte 5 is stable?) to do the following: 

1. to create a new version that ports this to  Svelte 5 (including plugins, use Runes and remove deprecated features that will be removed in Svelte 6) 

2. to do some research and write from guiding instructions on how to theoretically use the `Carta` object and build our the vanilla js components for the following:

- [X] `MarkdownEditor`
- [X] `Markdown`
- [X] `Toolbar`
- [X] `Input`
- [X] `Renderer`

 I can imagine this library being used in a lot of projects, and frameworks, and I'm willing to put work to port this to vanilla JS (and possibly Vue and React as well)... But I need some guidance (I haven't used Svelte 4 before and I'm quite confused about the architecture system design decisions and responsibilities of the components... ), 

I'm imagining it will be something like [quill mention](https://github.com/quill-mention/quill-mention/blob/master/docs/index.html#L149
) or [Niivue](https://github.com/quill-mention/quill-mention/blob/master/docs/index.html#L149) 


Where we can create our own components like (Just throwing initial ideas around)

```js
const carta = new Carta({ extensions: [hashtag()] });

const new MarkdownEditor(
  { carta,
    debounce: 500,
    onDebouncedChange: () => {
      // maybe get the AST here
      // carta.ASTRepresentation 
      // so we can save it on our server
      // or plug it in on our own previewer 
    },
    onDispatchEvent: (eventType, payload) => {
      // maybe do something given eventType and payload
    }
  }).attachToId("carta-markdown-editor-container")

<div id="carta-markdown-editor-container" />

```

I know these tasks are quite involved, but I think you have done incredible work..

- [X]  striking a good balance between simplicity, and extensibility (I feel like milkdown is quite bloated with builtin prose mirror)...  
- [X] I love that you have build awesome plugins as well...  (which can serve
as as a good starting point to build our own plugins)... 
- [X] awesome documentation on how to create advance extensions...
- [X] I can go on and on on how amazing this project is. 🤩 ❤️ 

 I'm willing to put my money where my mouth by personally sponsoring you for the two tasks if I can afford it... 

 Again.. amazing project. Hopefully you can give me a number (in `USD`, we can talk about this privately through email), and I'll save up 🙂.
 
 Looking forward to hear back from you 😀

**Comment by BearToCode** [comment]
Hi @mithi,

I appreciate very much your comment 😄 . I don't think I'll be able to work on this for at least a couple of months, due to my enrollment in university, but I should be able to do something then.

I think you asked for two very different modifications in your comment in terms of work required: migrating to Svelte 5 and creating a vanilla JS version of the library.

For the first one, even thought Svelte 5 is not yet definitive and I've not worked directly with it, I'm quite confident not too much work will be needed. This project does not make extensive use of Svelte components, so migrating shoud be straightforward. I'd still wait for a stable release before doing so(as you said).

Providing a vanilla version of the library is a lot more tricky and requires planning. There are several points that need to be adressed to make it viable, as Svelte(Kit) does [not provide](https://github.com/sveltejs/kit/issues/2572) a simple universal way to do so:
- Should we provide web components (`<markdown-editor>`) or a class-component to be mounted (`new MarkdownEditor`)? I've personally compiled another project of mine to web components and it was no easy task. I'd prefer to use a vanilla class, but I'll have to check out how to do that...
- How are embedded Svelte styles handled?
- Will custom elements provided by plugins, which are embedded in the editor (e.g. the search features of the Slash and Emoji plugins) still work? Probably all options that offer `SvelteComponent`s won't be available.
- Make sure all the plugins will still work.
- Will SSR work?
- Hope that everything will work fine with Svelte 5 🤞 

There are probably other things that I forgot/will discover during developing.

Regarding the _commission_: I've created this project because I needed an editor myself, and I decided to share it. I don't see developing OSS as a job. For this reason, I don't quite like the idea of a commision.
Still, if you find this project valuable and plan to make a _donation_, based on how much value it has provided you, I'll highly appreciate it ❤️ .
If you need the previously mentioned support to be able to use the editor, maybe wait until it has been implemented. I might look into setting up a GitHub sponsor page.


If in the meantime I hopefully get time to dedicate to this, you'd like to start working on it yourself, start making some other changes or have any question regarding the project, feel free to ask.

**Comment by mithi** [comment]
Hi @BearToCode 

I really appreciate your detailed response to my query. 

Sorry I think I did not word the second request properly,
I didn't mean to ask you do create a vanilla JS version of the library, that would be so too much work for you!

It's more like I noticed that the Carta class is purely written  in typescript and I thought that it would be possible for for me
the use that Carta instance and build my own Markdown editor to use this class instead (in any framework I prefer).

I was just hoping I could commision you to write more detailed documentation on how the svelte component `MarkdownEditor` (and other surrounding components like `Input`) uses the Carta instance under the hood...

A more detailed documentation to help provide some direction and guidance so I can write my own `MarkdownEditor` in other frameworks  - a `MarkdownEditor` Component (in React and/or Vue) that uses the `Carta` instance you have written. 

I tried to read your `MarkdownEditor` component (and the other components surrounding it like the `InputEnhancer` and `Renderer`) but it's quite difficult for me to grasp (partially because I haven't played around with Svelte 4 syntax as much as I should), and also unfortunately I'm also a bit overwhelmed since I have very little understanding on the system design of this library. 

I can try to spend a bit more time to try to explore and play around but then I thought maybe it's more efficient to just ask you directly for help since you know this library inside out (since you wrote it!)

I also see as you've mentioned that this library does not make extensive use of Svelte components, so I was thinking that maybe I can actually migrate it myself (I understand you're very busy), as Svelte 5 is now on `Release Candidate` phase (https://svelte.dev/blog/svelte-5-release-candidate)

I quote from there blog post:

> It doesn't mean that it's ready for production, or that nothing will change between now and 5.0. But if you've held off on dabbling with Svelte 5 during the public beta phase, now is a great time to try it out...

**So TLDR, I was wondering if you can I can ask you to write more detailed documentation on how the Svelte 5 components you wrote utilizes the Carta object so I can rewrite components in other frameworks to use the Carta object (and Carta plugins and extensions). The documentation would also help me migrate your existing components (including some of your plugins like the `Slash` one), from Svelte 4 to Svelte 5, if you don't have time to do so. And if so, it would be nice to discuss this privately on how maybe I can provide some compensation for this effort to write this documentation.** 

Again, thank you so much for creating this library and making it open source.. It's the most awesome extendable markdown editor I've seen ever! 🥳 🥳  👏  👏

**Comment by mithi** [comment]
Hi @BearToCode 

I really appreciate your detailed response to my query. 

Sorry I think I did not word the second request properly, I didn't mean to ask you do create a vanilla JS version of the library, that would be so too much work for you!

It's more like I noticed that the Carta class is purely written  in typescript and I thought that it would be possible for for me
the use that Carta instance and build my own Markdown editor to use this class instead (in any framework I prefer).

I was just hoping I could request to you to write more detailed documentation on how the svelte component `MarkdownEditor` (and other surrounding components like `Input`) uses the Carta instance under the hood...

I understand that you don't quite like the idea of commision, then I'd gladly make a donation (Maybe through Paypal or Buy me a Coffee (Like https://ko-fi.com/minimithi) if you do have time to write it. 

A more detailed documentation to help provide some direction and guidance so I can write my own `MarkdownEditor` in other frameworks  - a `MarkdownEditor` Component (in React and/or Vue) that uses the `Carta` instance you have written. 

I tried to read your `MarkdownEditor` component (and the other components surrounding it like the `InputEnhancer` and `Renderer`) but it's quite difficult for me to grasp (partially because I haven't played around with Svelte 4 syntax as much as I should), and also unfortunately I'm also a bit overwhelmed since I have very little understanding on the system design of this library. 

I can try to spend a bit more time to try to explore and play around but then I thought maybe it's more efficient to just ask you directly for help since you know this library inside out (since you wrote it!)

I also see as you've mentioned that this library does not make extensive use of Svelte components, so I was thinking that maybe I can actually migrate it myself (I understand you're very busy), as Svelte 5 is now on `Release Candidate` phase (https://svelte.dev/blog/svelte-5-release-candidate)

I quote from there blog post:

> It doesn't mean that it's ready for production, or that nothing will change between now and 5.0. But if you've held off on dabbling with Svelte 5 during the public beta phase, now is a great time to try it out...

**So TLDR, I was wondering if you can I can ask you to write more detailed documentation on how the Svelte 5 components you wrote utilizes the Carta object so I can rewrite components in other frameworks to use the Carta object (and Carta plugins and extensions). The documentation would also help me migrate your existing components (including some of your plugins like the `Slash` one), from Svelte 4 to Svelte 5, if you don't have time to do so. I understand that you don't quite like the idea of commission, then I'd gladly make a donation  if you do have time to write it.** 

Again, thank you so much for creating this library and making it open source.. It's the most awesome extendable markdown editor I've seen ever! 🥳 🥳  👏  👏

**Comment by mithi** [comment]
Hi @BearToCode 

I really appreciate your detailed response to my query. 

I'm sorry I think I did not word the second request properly, I didn't mean to ask you do create a vanilla JS version of the library, that would be too much work for you!

It's more like I noticed that the Carta class is purely written in Typescript and I think that it would be possible for for me
to use that Carta instance and build my own Markdown editor on top of it instead (can be in Svelte 5 or VueJS or React).

I was just hoping I could request to you to write more detailed documentation on how the svelte 4 component `MarkdownEditor` (and other surrounding components like `Input`) uses the Carta instance under the hood...

I believe that a documentation will help provide some direction and guidance so I can write my own `MarkdownEditor` in other frameworks  - a `MarkdownEditor` Component (in React and/or Vue) that uses the `Carta` instance you have written. 

I also see as you've mentioned that this library does not make extensive use of Svelte components, so I was thinking that maybe I can actually migrate it myself, if you don't have time to do so. Svelte 5 is now on `Release Candidate` phase (https://svelte.dev/blog/svelte-5-release-candidate)

I quote from there blog post:

> It doesn't mean that it's ready for production, or that nothing will change between now and 5.0. But if you've held off on dabbling with Svelte 5 during the public beta phase, now is a great time to try it out...

I understand that you don't quite like the idea of commission, then I'd gladly make a donation (Maybe through PayPal or Buy me a Coffee (Like https://ko-fi.com/minimithi) if you do have time to write this.

I tried to read your `MarkdownEditor` component (and the other components surrounding it like the `Input` and `Renderer`) but it's quite difficult for me to grasp (partially because I haven't played around with Svelte 4 as much as I should), and also unfortunately I'm also a bit overwhelmed since I have very little understanding on the system design of this library, and the reasonings behind design decisions. 

I can try to spend a bit more time to try to explore and play around but then I thought maybe it's more efficient to just ask you directly for help since you know this library inside out (you wrote it!)

# TLDR

> **I was wondering if you can I can ask you to write more detailed documentation on how the Svelte 4 components you wrote utilizes the Carta object so I can build a custom markdown editor 
in other frameworks to use the Carta object. The documentation would also help me migrate your existing components (including some of your plugins like the `Slash` one), from Svelte 4 to Svelte 5, if you don't have time to do so. I understand that you don't quite like the idea of commission, then I'd gladly make a donation if you do have time to write it.** 

Again, thank you so much for creating this library and making it open source.. It's the most awesome extendable markdown editor I've seen ever! 🥳 🥳  👏  👏

**Comment by mithi** [comment]
Hi  @BearToCode 

I really appreciate your detailed response to my query. 

I'm sorry I think I did not word the second request properly, I didn't mean to ask you do create a vanilla JS version of the library, that would be too much work for you!

It's more like I noticed that the Carta class is purely written in Typescript and I think that it would be possible for for me
to use that Carta instance and build my own Markdown editor on top of it instead (can be in Svelte 5 or VueJS or React).

I was just hoping I could request to you to write more detailed documentation on how the svelte 4 component `MarkdownEditor` (and other surrounding components like `Input`) uses the Carta instance under the hood...

I believe that a documentation will help provide some direction and guidance so I can write my own `MarkdownEditor` in other frameworks  - a `MarkdownEditor` Component (in React and/or Vue) that uses the `Carta` instance you have written. 

As you've mentioned that this library does not make extensive use of Svelte components, so I was thinking that maybe I can actually migrate it myself, if you don't have time to do so. Svelte 5 is now on `Release Candidate` phase (https://svelte.dev/blog/svelte-5-release-candidate)

I quote from there blog post:

> It doesn't mean that it's ready for production, or that nothing will change between now and 5.0. But if you've held off on dabbling with Svelte 5 during the public beta phase, now is a great time to try it out...

I understand that you don't quite like the idea of commission, I'd gladly make a donation (Maybe through PayPal or Buy me a Coffee (Like https://ko-fi.com/minimithi) as an appreciation of your time and effort into writing this.

I tried to read your `MarkdownEditor` component (and the other components surrounding it like the `Input` and `Renderer`) but it's quite difficult for me to grasp (partially because I haven't played around with Svelte 4 as much as I should), and also unfortunately I'm also a bit overwhelmed since I have very little understanding on the system design of this library, and the reasonings behind design decisions. 

I can try to spend a bit more time to try to explore and play around but then I thought maybe it's more efficient to just ask you directly for help since you know this library inside out (naturally because you wrote it!)

# TLDR

> **I was wondering if you can I can ask you to write more detailed documentation on how the Svelte 4 components you wrote utilizes the Carta object under the hood so I can build my custom markdown editor 
in other frameworks (and in Svelte 5) to use the Carta object. The documentation would also help me migrate your existing components (including some of your plugins like the `Slash` one), from Svelte 4 to Svelte 5, if you don't have time to do so. I understand that you don't quite like the idea of commission, I'd gladly make a donation as an appreciation of your time and effort into writing this.** 

Again, thank you so much for creating this library and making it open source... It's the most awesome extendable markdown editor I've seen ever! 🥳 🥳 👏 👏

**Comment by BearToCode** [comment]
Hi again,

USING the `Carta` inside to build "components" for a vanilla version or other framerworks seems like a great idea! It probably has some limitations, but overall it is much easier and more robust than just compiling the components.

If you just need me to explain the roles of the components and how they work, I can do that easily. I think I could add an header to the various components explaining them, and add descriptions to their functions and properties:

```svelte
<!-- 
	@description A description of the what the component does and how it does that
-->
<script lang="ts">
  /**
   * A description of the property.
   */
  export let prop: Something;
  
  /**
   * What the function does
   * @param a What is a
   * @param b What is b
   */
  function fn(a, b) {
  
  }
</script>
```

Let me know if something like this would be enough, or if you had another idea in mind. You could obviously contact for any question.


Following your suggestion I create a [ko-fi page](https://ko-fi.com/beartocode).

**Comment by mithi** [comment]
>  I think I could add an header to the various components explaining them, and add descriptions to their functions and properties

Yes that would be really cool! But if you're doing this just for me, and if it will take too much time from more important things, then please don't prioritize this! After familiarizing myself with Svelte more the past few hours I think 
I can read and understand the code a bit better and I'll message you if I have any questions or clarifications. 

I'm hoping to contribute more to this project in the near future when I'm no longer swamped with work... So no hurry really... 

> Following your suggestion I created a [ko-fi page](https://ko-fi.com/beartocode).

I have made a little donation on your Ko-Fi page, and I'll be making more donations in the future once I get to spend more time playing with this library. 

Again, thank you so much for entertaining all my questions! 🙂 

I will message you again in the next couple of months 🙂

**Comment by BearToCode** [comment]
Thanks a lot for your donation! Looking forward to working with you 😄

**Comment by warren-bank** [comment]
just a quick drive-by comment regarding this issue..

comparison of this project to another project that uses Svelte and publishes a vanilla JS library:

<details>
  <summary>carta-md</summary>
  
  * [jsdelivr: carta-md](https://www.jsdelivr.com/package/npm/carta-md)
  * [jsdelivr: carta-md/dist](https://www.jsdelivr.com/package/npm/carta-md?tab=files&path=dist)
  * [jsdelivr: carta-md/dist/index.js](https://cdn.jsdelivr.net/npm/carta-md@4.2.1/dist/index.js)
    - npm package includes raw _.svelte_ files
  * [github: package.json/scripts.build](https://github.com/BearToCode/carta/blob/carta-md-v4.2.1/package.json#L15)
    - `npm run build` calls: `node scripts/build.js`
  * [github: scripts/build.js](https://github.com/BearToCode/carta/blob/carta-md-v4.2.1/scripts/build.js)
    - foreach directory in _packages_, call: `npm run build`
  * [github: packages/carta-md/package.json/scripts.build](https://github.com/BearToCode/carta/blob/carta-md-v4.2.1/packages/carta-md/package.json#L22)
    - `npm run build` calls: `svelte-kit sync && svelte-package`
</details>

<details>
  <summary>bigger-picture</summary>

  * [github: README#install](https://github.com/henrygd/bigger-picture/tree/v1.1.17#install)
  * [jsdelivr: bigger-picture](https://www.jsdelivr.com/package/npm/bigger-picture)
  * [jsdelivr: bigger-picture/dist](https://www.jsdelivr.com/package/npm/bigger-picture?tab=files&path=dist)
  * [jsdelivr: bigger-picture/dist/bigger-picture.min.js](https://cdn.jsdelivr.net/npm/bigger-picture@1.1.17/dist/bigger-picture.min.js)
    - npm package only includes transpiled javascript
  * [github: package.json/scripts.build](https://github.com/henrygd/bigger-picture/blob/v1.1.17/package.json#L60-L68)
    - build scripts use `rollup` (and `sass`)
  * [github: package.json/devDependencies](https://github.com/henrygd/bigger-picture/blob/v1.1.17/package.json#L70-L84)
    - dev dependencies include several `rollup` plugins, including one specifically for _Svelte_
</details>

summary:
* `bigger-picture` is a front-end library written in _Svelte_ that compiles to vanilla JS using `rollup` as a build tool

assertion:
* the same could be done for `carta-md`
* if the _&#64;cartamd/_ plugins don't include any _Svelte_ components, their current npm packages could probably be used as-is

**Comment by warren-bank** [comment]
just for fun.. I spent a little time this evening attempting to make a browser build.
early days, but the results are looking pretty successful.
links..

* [github: directory](https://github.com/warren-bank/fork-js-svelte-carta/tree/workspace/3-browser-builds)
* [github: test build](https://github.com/warren-bank/fork-js-svelte-carta/tree/workspace/3-browser-builds/02-upstream-master)
* [github: html demo](https://github.com/warren-bank/fork-js-svelte-carta/blob/workspace/3-browser-builds/02-upstream-master/test/index.html)
* [githack: html demo](https://rawcdn.githack.com/warren-bank/fork-js-svelte-carta/3b18706ce90c59be8473c249067ceeede866dd70/3-browser-builds/02-upstream-master/test/index.html)
  - using [githack](https://raw.githack.com/) CDN to view the page since my branch isn't configured to use Github Pages

**Comment by warren-bank** [comment]
refreshed browser build..
* from [commit 088274f](https://github.com/BearToCode/carta/commit/088274f7b3eae81c78d7623dc6a764b241089eb1)
  - even with tag [@cartamd/plugin-tikz-v4.1.6](https://github.com/BearToCode/carta/releases/tag/%40cartamd%2Fplugin-tikz-v4.1.6)
  - using [builds](https://github.com/warren-bank/fork-js-svelte-carta/tree/be7d6a0010a777e4785ab86cf00864eb3f5b7c9e/1-builds/07-upstream-master/packages)
* [github: html demo code](https://github.com/warren-bank/fork-js-svelte-carta/blob/be7d6a0010a777e4785ab86cf00864eb3f5b7c9e/3-browser-builds/07-upstream-master/test/index.html)
  - [githack: html demo](https://rawcdn.githack.com/warren-bank/fork-js-svelte-carta/be7d6a0010a777e4785ab86cf00864eb3f5b7c9e/3-browser-builds/07-upstream-master/test/index.html)
